### ISSUE TITLE
[clang] `error: alias must point to a defined variable or function` false negative when optimizations are enabled

### ISSUE THREAD

**Issue by nickdesaulniers** [llvm/llvm-project#89474]
https://godbolt.org/z/TrGeazYc5

```c
extern inline __attribute__((gnu_inline))
void bar (void) {}

void foo [[gnu::alias("_Z3barv")]] (void);
void foo2 (void) __attribute__((alias("_Z3barv")));
void foo3 (void) asm("_Z3barv");
```
<source>:4:12: error: alias must point to a defined variable or function
    4 | void foo [[gnu::alias("_Z3barv")]] (void);
      |            ^
<source>:4:12: note: the function or variable specified in an alias must refer to its mangled name
<source>:5:33: error: alias must point to a defined variable or function
    5 | void foo2 (void) __attribute__((alias("_Z3barv")));
      |                                 ^
<source>:5:33: note: the function or variable specified in an alias must refer to its mangled name
```
via #60481

**Comment by shafik** [comment]
Worth noting this crashes on asserts build: https://godbolt.org/z/basjzojxE

```console
Alias must point to a definition
ptr @_Z3foov
Alias must point to a definition
ptr @_Z4foo2v
fatal error: error in backend: Broken module found, compilation aborted!
```

**Comment by shafik** [comment]
It looks like we diagnose this in codegen:

https://github.com/llvm/llvm-project/blob/adc4f6233df734fbe3793118ecc89d3584e0c90f/clang/lib/CodeGen/CodeGenModule.cpp#L591-L594

and:

https://github.com/llvm/llvm-project/blob/adc4f6233df734fbe3793118ecc89d3584e0c90f/clang/lib/CodeGen/CodeGenModule.cpp#L612-L619

I think then it makes more sense now that this is optimization sensitive.

CC @AaronBallman @erichkeane

**Comment by AaronBallman** [comment]
Both of those checks should always be called: https://github.com/llvm/llvm-project/blob/dc20a0ea1fd04a2ef95eb5c73e9f88357fc5f694/clang/lib/CodeGen/CodeGenModule.cpp#L832 calls `checkAliases()` which has no optimization-specific logic. So perhaps optimizations are removing some aliases from the list?
